src/build/app/mod.rs

```diff
@@ -2416,62 +2443,95 @@ impl<'help> App<'help> {
         }
     }
 
-    pub(crate) fn _create_help_and_version(&mut self) {
-        debug!("App::_create_help_and_version");
+    #[allow(clippy::blocks_in_if_conditions)]
+    pub(crate) fn _check_help_and_version(&mut self) {
+        debug!("App::_check_help_and_version");
 
-        if !(self.is_set(AppSettings::DisableHelpFlag)
-            || self
-                .args
-                .args()
-                .any(|x| x.long == Some("help") || x.id == Id::help_hash())
+        if self.is_set(AppSettings::DisableHelpFlag)
+            || self.args.args().any(|x| {
+                x.provider == ArgProvider::User
+                    && (x.long == Some("help") || x.id == Id::help_hash())
+            })
             || self
                 .subcommands
                 .iter()
-                .any(|sc| sc.long_flag == Some("help")))
+                .any(|sc| sc.long_flag == Some("help"))
         {
-            debug!("App::_create_help_and_version: Building --help");
-            let mut help = Arg::new("help")
-                .long("help")
-                .about(self.help_about.unwrap_or("Prints help information"));
+            debug!("App::_check_help_and_version: Removing generated help");
 
-            if !(self.args.args().any(|x| x.short == Some('h'))
+            let generated_help_pos = self
+                .args
+                .args()
+                .position(|x| x.id == Id::help_hash() && x.provider == ArgProvider::Generated);
+
+            if let Some(index) = generated_help_pos {
+                self.args.remove(index);
+            }
+        } else {
+            let other_arg_has_short = self.args.args().any(|x| x.short == Some('h'));
+            let help = self
+                .args
+                .args_mut()
+                .find(|x| x.id == Id::help_hash())
+                .expect(INTERNAL_ERROR_MSG);
+
+            if !(help.short.is_some()
+                || other_arg_has_short
```

Comment:
Move the `self.args.args().any(|x| x.short == Some('h'))` down here. Can prevent some unnecessary computation when the first condition is true.

Reply:
Can't do that because `help` uses `args_mut`.